File: _downloads/2313115342724c737fadc1a2f989ca0d/ihw5_6.m
function ihw5_6()
% Ingle/Proakis HW Problem P5.6: 
% Compares DTFT to DFT for a sequence.
%
% Copyright 1998 Cameron H. G. Wright 
%
x = [1,2,3,4,5,6,6,5,4,3,2,1]; % data sequence x[n] 
% optional variation to play "whatif?"
% x = [0:0.5:6.5, 6:-0.5:0.5]; % modifed data sequence x[n] 
% x=[x, zeros(1,25)];  % zero padding of data sequence x[n]
n = 0:length(x)-1;
figure(1)
stem(n,x)
title('Discrete-Time Data Sequence'); 
xlabel('Sample'); 
ylabel('Magnitude')
% DFT analysis
kd=n;
Xd=fft(x);
magXd = abs(Xd); angXd = angle(Xd);
figure(2);
% plot with negative frequency swapped to the left half
% use normalized pi units for discrete radian frequency axis
wn= (kd-floor(length(kd)/2))/floor(length(kd)/2);
subplot(2,1,1); stem(wn,fftshift(magXd)); grid on;
title('DFT Magnitude'); 
xlabel('frequency in pi units'); 
ylabel('Magnitude')
subplot(2,1,2); stem(wn,fftshift(angXd)); grid on;
title('DFT Phase'); 
xlabel('frequency in pi units'); 
ylabel('Radians')

% DTFT analysis
kf = -500:500; w = (pi/500)*kf;  % [-pi, pi] axis divided into 1001 points.
Xf = x * (exp(-j*pi/500)) .^ (n'*kf); % DTFT using matrix-vector product
magXf = abs(Xf); angXf = angle(Xf);
figure(3)
subplot(2,1,1); plot(w/pi,magXf); grid
title('DTFT Magnitude'); 
xlabel('frequency in pi units'); 
ylabel('Magnitude')
subplot(2,1,2); plot(w/pi,angXf); grid
title('DTFT Angle'); 
xlabel('frequency in pi units'); 
ylabel('Radians')

% Superimpose the two plots
% Note the x-axis range must be the same for both
figure(4);
subplot(2,1,1); stem(wn,fftshift(magXd),'r');
hold on; plot(w/pi,magXf); 
title('DFT and DTFT Magnitude'); 
xlabel('frequency in pi units'); 
ylabel('Magnitude')
hold off;
subplot(2,1,2); stem(wn,fftshift(angXd),'r'); 
hold on; plot(w/pi,angXf);
title('DFT and DTFT Angle'); 
xlabel('frequency in pi units'); 
ylabel('Radians')
hold off;

figure(1)

File: _downloads/7169058ef43564b8a9cefbca805797bb/ihw5_6.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 14 15:22:26 2023

@author: george.york
"""
import matplotlib.pyplot as plt
import numpy as np
# Ingle/Proakis HW Problem P5.6: 
# Compares DTFT to DFT for a sequence.

# make a data sequence, a triangle
x = np.zeros(12)
value = 1.
for i in range(0, 6):
    x[i] = value
    value = value + 1
value = 6
for i in range(6, 12):
    x[i] = value
    value = value - 1

n = np.arange(0,len(x),1)  # goes from 0 to N-1, incrementing by 1

# plot the data
plt.figure(1)
plt.stem(n,x)
plt.title('Discrete-Time Data Sequence'); 
plt.xlabel('Sample'); 
plt.ylabel('Magnitude')   

# DFT analysis
kd = n
Xd = np.fft.fft(x)
magXd = np.abs(Xd)
angXd = np.angle(Xd)
# plot with negative frequency swapped to the left half
# use normalized pi units for discrete radian frequency axis 
plt.figure(2)

N = len(x)
wn = (n - 1 * (N/2)) / (N/2)  # shift axis for plot, and scale to range to pi (+/- 1)
plt.subplot(211)
plt.stem(wn, np.fft.fftshift(magXd))
plt.grid()
plt.title('DFT Magnitude'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Magnitude')
plt.subplot(212)
plt.stem(wn, np.fft.fftshift(angXd))
plt.grid()
plt.title('DFT Phase'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Radians')

# DTFT Analysis
# kf = np.arange(-500,501,1)  # [-pi, pi] axis divided into 1001 points.
# w = (np.pi/500)*kf;  
# kf_conj_transpose = np.matrix(kf)
# Xf = np.power(x * (np.exp(-1.j*np.pi/500)), (n*kf_conj_transpose.getH())) # DTFT using matrix-vector product
# Instead, use DFT with many samples to appear as if it is the DTFT (continious)

N2 = 600
n2 = np.arange(0,N2,1)  # goes from 0 to N-1, incrementing by 1
kd2 = n2
# zero pad out to 600 samples to fool FFT
Xd2 = np.fft.fft(x, N2)
magXd2 = np.abs(Xd2)
angXd2 = np.angle(Xd2)

plt.figure(3)

wn2 = (n2 - 1 * (N2/2)) / (N2/2)  # shift axis for plot, and scale to range to pi (+/- 1)
plt.subplot(211)
plt.plot(wn2, np.fft.fftshift(magXd2))
plt.grid()
plt.title('DTFT Magnitude'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Magnitude')
plt.subplot(212)
plt.plot(wn2, np.fft.fftshift(angXd2))
plt.grid()
plt.title('DTFT Phase'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Radians')

# Superimpose the two plots
# Note the x-axis range must be the same for both
plt.figure(4)
plt.subplot(211)    
plt.stem(wn, np.fft.fftshift(magXd), 'r')
#plt.hold(True)  
plt.plot(wn2, np.fft.fftshift(magXd2))
plt.title('DFT and DTFT Magnitude'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Magnitude')

#plt.hold(False)
plt.subplot(212)  
plt.stem(wn, np.fft.fftshift(angXd), 'r')
#plt.hold(True)   
plt.plot(wn2, np.fft.fftshift(angXd2))
plt.title('DFT and DTFT Angle'); 
plt.xlabel('frequency in pi units'); 
plt.ylabel('Radians')
#plt.hold(False)
plt.show()


